core/src/main/java/org/infinispan/util/concurrent/CommandAckCollector.java

```diff
@@ -0,0 +1,482 @@
+         }
+         if (this.topologyId != topologyId || !owners.remove(from)) {
+            //already received!
+            return;
+         }
+         this.returnValue = returnValue;
+
+         if (!success) {
+            //we are not receiving any backups ack!
+            owners.clear();
+            future.complete(returnValue);
+            if (trace) {
+               log.tracef("[Collector#%s] Ready! Command not succeed on primary.", id);
+            }
+         } else if (owners.isEmpty()) {
+            markReady();
+         }
+      }
+
+      public synchronized void backupAck(int topologyId, Address from) {
+         if (trace) {
+            log.tracef("[Collector#%s] Backup ACK. Address=%s, TopologyId=%s (expected=%s)",
+                  id, from, topologyId, this.topologyId);
+         }
+         if (this.topologyId == topologyId && owners.remove(from) && owners.isEmpty()) {
+            markReady();
+         }
+      }
+
+      @Override
+      public synchronized void completeExceptionally(Throwable throwable, int topologyId) {
+         if (trace) {
+            log.tracef(throwable, "[Collector#%s] completed exceptionally. TopologyId=%s (expected=%s)",
+                  id, topologyId, this.topologyId);
+         }
+         if (this.topologyId != topologyId) {
+            return;
+         }
+         doCompleteExceptionally(throwable);
+      }
+
+      private void markReady() {
+         if (trace) {
+            log.tracef("[Collector#%s] Ready! Return value=%ss.", id, returnValue);
+         }
+         future.complete(returnValue);
+      }
+
+      private void doCompleteExceptionally(Throwable throwable) {
+         owners.clear();
+         future.completeExceptionally(throwable);
+      }
+
+      @Override
+      public synchronized boolean hasPendingBackupAcks() {
+         return owners.size() > 1 || //at least one backup + primary address
+               //if one is missing, make sure that it isn't the primary
+               owners.size() == 1 && !primaryOwner.equals(owners.iterator().next());
+      }
+
+      @Override
+      public CompletableFuture<Object> getFuture() {
+         return future;
+      }
+
+      @Override
+      public synchronized void onMembersChange(Collection<Address> members) {
+         if (!members.contains(primaryOwner)) {
+            //primary owner left. throw OutdatedTopologyException to trigger a retry
+            if (trace) {
+               log.tracef("[Collector#%s] The Primary Owner left the cluster.", id);
+            }
+            doCompleteExceptionally(OutdatedTopologyException.getCachedInstance());
+         } else if (owners.retainAll(members) && owners.isEmpty()) {
+            if (trace) {
+               log.tracef("[Collector#%s] Some backups left the cluster.", id);
+            }
+            markReady();
+         }
+      }
+   }
+
+   private static class PutMapCollector implements Collector<Map<Object, Object>> {
```

Comment:
This should rather be `MultiKeyCollector` (or some better name) to accomodate functional commands as well. At least I hope it will work for func commands, too.
And also rename the methods above.